pcsx2/Recording/VirtualPad/VirtualPad.cpp

```diff
@@ -0,0 +1,438 @@
+
+	// Bind Window Events
+	Bind(wxEVT_MOVE, &VirtualPad::OnMoveAround, this);
+	Bind(wxEVT_CLOSE_WINDOW, &VirtualPad::OnClose, this);
+	Bind(wxEVT_ICONIZE, &VirtualPad::OnIconize, this);
+	Bind(wxEVT_ERASE_BACKGROUND, &VirtualPad::OnEraseBackground, this);
+	// Temporary Paint event handler so the window displays properly before the controller-interrupt routine takes over with manual drawing.
+	// The reason for this is in order to minimize the performance impact, we need total control over when render is called
+	// Windows redraws the window _alot_ otherwise which leads to major performance problems (when GS is using the software renderer)
+	Bind(wxEVT_PAINT, &VirtualPad::OnPaint, this);
+	// DevCon.WriteLn("Paint Event Bound");
+
+	// Finalize layout
+	SetIcons(wxGetApp().GetIconBundle());
+	SetTitle(wxString::Format("Virtual Pad - Port %d", controllerPort + 1));
+	SetPosition(options.VirtualPadPosition);
+	SetBackgroundColour(*wxWHITE);
+	SetBackgroundStyle(wxBG_STYLE_PAINT);
+	// This window does not allow for resizing for sake of simplicity: all images are scaled initially and stored, ready to be rendered
+	SetWindowStyle(wxDEFAULT_FRAME_STYLE & ~wxRESIZE_BORDER & ~wxMAXIMIZE_BOX);
+
+	// Causes flickering, despite it supposed to be preventing it!
+	// SetDoubleBuffered(true);
+}
+
+void VirtualPad::OnMoveAround(wxMoveEvent& event)
+{
+	if (IsBeingDeleted() || !IsVisible() || IsIconized())
+		return;
+
+	if (!IsMaximized())
+		options.VirtualPadPosition = GetPosition();
+	event.Skip();
+}
+
+void VirtualPad::OnClose(wxCloseEvent& event)
+{
+	// Re-bind the Paint event in case this is due to a game being opened/closed
+	manualRedrawMode = false;
+	Bind(wxEVT_PAINT, &VirtualPad::OnPaint, this);
+	// DevCon.WriteLn("Paint Event Bound");
+	Hide();
+}
+
+void VirtualPad::OnIconize(wxIconizeEvent& event)
+{
+	if (event.IsIconized())
+	{
+		manualRedrawMode = false;
+		Bind(wxEVT_PAINT, &VirtualPad::OnPaint, this);
+		// DevCon.WriteLn("Paint Event Bound");
+	}
+}
+
+void VirtualPad::OnEraseBackground(wxEraseEvent& event)
+{
+	// Intentionally Empty
+	// See - https://wiki.wxwidgets.org/Flicker-Free_Drawing
+}
+
+void VirtualPad::OnPaint(wxPaintEvent& event)
+{
+	// DevCon.WriteLn("Paint Event Called");
+	wxBufferedPaintDC dc(this, wxBUFFER_VIRTUAL_AREA);
+	Render(dc);
+}
+
+void VirtualPad::Redraw()
+{
+	wxClientDC cdc(this);
+	wxBufferedDC dc(&cdc);
+	Render(dc);
+}
+
+void VirtualPad::Render(wxDC& bdc)
+{
+	// Update GUI Elements and figure out what needs to be rendered
+	for (VirtualPadElement* virtualPadElement : virtualPadElements)
+		virtualPadElement->UpdateGuiElement(renderQueue, clearScreenRequired);
+
+	// Update Graphic Elements off render stack
+	// Before we start rendering (if we have to) clear and re-draw the background
+	if (!manualRedrawMode || clearScreenRequired || !renderQueue.empty())
+	{
+		bdc.SetBrush(*wxRED);
+		bdc.DrawRectangle(wxPoint(0, 0), bdc.GetSize());
+		bdc.SetBrush(wxNullBrush);
```

Comment:
I do wonder if this is 100% necessary. Since the bitmap stretches over the entire window to begin with, I feel like this would just get covered up and may even cause a weird flicker of red in certain situations.

Reply:
It was used for testing purposes to see drawing artifacts from what i gathered.

Reply:
Ah i forgot to change this to white, which would alleviate the flicker concern.  It's been a long time since I originally wrote this code, I believe properly clearing the screen solved some issues, in general I think it's just a best practice.

Reply:
Fair